include/dem/load_balancing.h

```diff
@@ -0,0 +1,398 @@
+   * weight
+   *
+   * @param[in] cell The cell for which the load is calculated
+   * @param[in] status The status of the cell related to the coarsening level
+   *
+   * @return The total weight of the cell
+   */
+  unsigned int
+  calculate_total_cell_weight_with_mobility_status(
+    const typename parallel::distributed::Triangulation<dim>::cell_iterator
+                    &cell,
+    const CellStatus status) const;
+
+  /**
+   * @brief The load balancing method chosen by the user.
+   */
+  Parameters::Lagrangian::ModelParameters::LoadBalanceMethod
+    load_balance_method;
+
+  /**
+   * @brief The load balancing iteration check function according to the load
+   * balancing method.
+   */
+  std::function<bool()> iteration_check_function;
+
+  /**
+   * @brief Default hard-coded load weight of a cell.
+   */
+  const unsigned int cell_weight = 1000;
+
+  /**
+   * @brief Load weight of a particle, the default parameters is 10000.
+   */
+  unsigned int particle_weight;
+
+  /**
+   * @brief Load weight factor of particle weight in a cell with an inactive
+   * mobility status (only with ASC).
+   */
+  double inactive_status_factor;
+
+  /**
+   * @brief Load weight factor of particle weight in a cell with an active
+   * mobility status (only with ASC).
+   */
+  double active_status_factor;
+
+  /**
+   * @brief Iteration number for load balancing execution, only when method
+   * is `once`.
+   */
+  unsigned int load_balance_step;
+
+  /**
+   * @brief Frequency of load balancing execution, only when method is
+   * `frequent`.
+   */
+  unsigned int load_balance_frequency;
+
+  /**
+   * @brief Frequency of check for the load balancing execution, only when
+   * method is `dynamic`.
+   */
+  unsigned int dynamic_check_frequency;
+
+  /**
+   * @brief Threshold (\beta) for the check load balancing execution, only when
+   * method is `dynamic`.
+   *
+   * Load balancing will be done when the computational load amongst core is too
+   * uneven. If L_{max} - L_{min} > \beta \bar{L}, where L_{max}, L_{min}, and
+   * \bar{L} are the maximum, minimum, and the average load, the load balancing
+   * will be executed.
+   */
+  double load_threshold;
+
+  /**
+   * @brief MPI communicator.
+   */
+  MPI_Comm mpi_communicator;
+
+  /**
+   * @brief Number of MPI processes.
+   */
+  const unsigned int n_mpi_processes;
+
+  /**
+   * @brief Rank of the current MPI process.
+   */
+  const unsigned int this_mpi_process;
+
+  /**
+   * @brief Pointer to the simulation control object.
+   */
+  std::shared_ptr<SimulationControl> simulation_control;
+
+  /**
+   * @brief Pointer to the triangulation object.
+   */
+  parallel::distributed::Triangulation<dim> *triangulation;
+
+  /**
+   * @brief Pointer to the particle handler object.
+   */
+  Particles::ParticleHandler<dim> *particle_handler;
+
+  /**
+   * @brief Pointer to the adaptive sparse contacts object.
+   */
+  AdaptiveSparseContacts<dim> *adaptive_sparse_contacts;
```

Comment:
Any reasons why you are keeping raw points for some stuff and shared_ptr for others?
I presume the one for which you are keeping raw points are not pointer in the DEM class, is that it?

Reply:
Yes